Add/tld/fr extra fields dialog

deBhal opened the pull request:
This PR adds a form to support .FR extra fields for the for the domains checkout flow and contact detail editing flow as per #13890.

I've added the `domains/cctlds` feature flag, and the .FR form is behind that, so the most critical testing is that we don't break calypso with the flag off:
`DISABLE_FEATURES=domains/cctlds make run`

*Note*: that the back-end will only offer .FR domains to proxied a12s, or sandboxes with the store sandbox enabled, so this is not a very community friendly PR, sorry :( 

After that, we want to see:
- the checkout flow works: contact details -> .fr details (for .fr domains) -> maybe privacy dialog -> checkout
- The appropriate fields being visible for registrant type and country of birth
- the (token) tests run (and don't break the entire suite anymore): ` npm run test-client client/components/domains/registrant-extra-info` (or even better  `npm run test-client`)
- the extra fields are passed to the server

Things that are not in this PR:
- no sanitation or validation at the moment ( we'll do that after we reduxify)
- no registration success (we need to fix the backend)
- no remembered data on the .fr form ( needs backend + redux )
- domain management integration (after redux)
- more testing (after redux)

Video:
![fr extra 2](https://cloud.githubusercontent.com/assets/5952255/26438436/637bbea4-4166-11e7-97fe-7180638f779d.gif)

And passing that data to the server:
![thank_you_ _wordpress_com_and_licecap_v1_25__stopped__and_calypso_ _-bash_ _86x85](https://cloud.githubusercontent.com/assets/5952255/26438439/668bd0b6-4166-11e7-9835-d34c188ded79.jpg)

umurkontaci commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -56,7 +59,8 @@ export default React.createClass( {
 			submissionCount: 0,
-			phoneCountryCode: 'US'
+			phoneCountryCode: 'US',
+			registrantExtraInfo: false
```
Is this always boolean? If so, do you think it'd make it easier to understand if we prefix it with `has` or `needs`?

umurkontaci commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -212,8 +227,25 @@ export default React.createClass( {
+
+	handleFrSubmit( registrantExtraInfo ) {
+		this.setState( { registrantExtraInfo } );
+		this.closeDialog( 'fr' );
```
I think we never open a dialog named `fr`. Did you mean `isFrDialogVisible`?
I think it'd would make it more resilient to just create functions for every dialog (or use const) instead of relying on literal strings sprinkled everywhere.

deBhal commented:
@mikeshelton1503: No interaction yet, but I think that screenshot implements your design if you wanted to have a look and give me design feedback, especially around the DOB.

deBhal commented:
@umurkontaci: Thanks for catching the `fr` typo, but I think I need to rethink the logic there anyway. `needsExtraRegistrantInfo()` isn't going to work once we pre-fill data or add a back button, so I think what I'll need to do is have matching arrays of "needsDialogX:bool" (based on props) and "hasSubmittedDialogX: bool" (based on state).

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,228 @@
+		return (
+			<div>
+				<Card className="registrant-extra-info__title-card">
+						{ this.props.translate( '.FR Registration' ) }
```
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
Ok, I've reworked the extras form to be thinner, and moved it out of the dialog and into the details form.

I've also added some "steps" logic, which is probably re-inventing a wheel, but might nonetheless be useful in #12622 if we wanted to add the privacy dialog in as another step. I've updated the screenshot to a video of the flow. We definitely need a back button.

I think I've settled on how we should handle the extra details forms too, as described in the README. If we want them to be usable in both the domain details and checkouts, we need them to be super-thin, so what I'm going for now is just displaying some fields with error messages, and passing out the value in all the fields to the parent, and leaving validation, sanitation etc to them.

It should be reasonably clean once we get the contact details out of the react state and into the redux state.

yoavf commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,216 @@
+						checked={ 'individual' === this.state.registrantType }
+						name="registrantType"
+						onChange={ this.handleChangeEvent } />
+					<span>An individual</span>
```
Missing translate (a few additional occurrences below)

yoavf commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,216 @@
+				<FormFieldset>
+					<FormLabel>
+						{ this.props.translate( "Who's this domain for?" ) }
+					</FormLabel>
```
Closing FormLabel should be at the end of the FormRadio element

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,223 @@
+							checked={ 'individual' === this.state.registrantType }
+							name="registrantType"
+							onChange={ this.handleChangeEvent } />
+						<span>{ this.props.translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
There's no way to translate the "content" field in a css "After" rule, is there?
 ( https://github.com/Automattic/wp-calypso/pull/14156/commits/047a536b24761ee8dedcf7a7223e716f8b0574b2 )

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,273 @@
+							name="registrantType"
+							checked={ 'individual' === this.state.registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ this.props.translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
Update: It's alive!

All the fr-form fields are hooked up to the form's state, and in turn to contact details, which are then passed on to the purchase and validate endpoints, huzzah!:

![merci_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/26341210/ec28287e-3fd5-11e7-9b71-16f19111d92b.jpg)

It doesn't actually work yet because the backend just ignores the extra fields right now, but we're nearly there!

Validation and sanitation don't work, and there is some unpleasant hardwiring in there, but I'd like to punt that to a later PR and solve it by reduxifying everything. Showing different forms and passing values and errors back and forwards is going to be so much easier with the data in a redux state.

It could use another look over, but It's getting late, so I'm going to mark this as "for review" now, just be on the lookout for stupidity.

deBhal commented:
On the design front, I Italicized the 'Option', and the date entry needs a look

blowery commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -415,7 +490,16 @@ class DomainDetailsForm extends Component {
+		}
+	}
+
+	render = () => {
```
render doesn't need to be bound to the instance

blowery commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -330,24 +391,38 @@ class DomainDetailsForm extends Component {
 
-	closeDialog = () => {
-		this.setState( { isDialogVisible: false } );
+	closeDialog( dialogKey = 'isDialogVisible' ) {
```
these _might_ need to be bound to the instance like they were before, since they're using `this`. just depends on how they're called..

deBhal replied:
Good catch! They do need to be bound to work when passed in to the `PrivacyProtection` component [here](https://github.com/Automattic/wp-calypso/blob/68273016cb2e6581a807f23bb08788cdb3a65c4b/client/my-sites/upgrades/checkout/domain-details-form.jsx#L267)

blowery commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -53,15 +61,23 @@ class DomainDetailsForm extends Component {
 	}
 
-	componentWillMount() {
+	componentWillMount = () => {
```
this doesn't need to be bound

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,301 @@
+		this.setState( { [ event.target.name ]: event.target.value } );
+	}
+
+	shouldComponentUpdate = ( nextProps, nextState ) => {
```
I think `React.PureComponent` does basically the same :)

deBhal replied:
Nearly, but not quite :( `isEqual()` is a deep comparison, and PureComponent uses a shallow one https://facebook.github.io/react/docs/react-api.html#react.purecomponent.

I had to add this to break an infinite rerender caused by passing the state up out of this component and adding it to the state of it's parent. This was the point at which I decided that I had forced this too far, and need to stop and reduxify it all :)

klimeryk replied:
Ah, didn't realize it was deep comparison 👍

deBhal commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,301 @@
+			this.props.onStateChange( this.state );
+	}
+
+	getRelevantFields( state ) {
```
> @hewsut: what does `getRelevantFields` do? I mean, I understand what it does inside the function. But I don’t see it called anywhere?

Nice catch! It captures the way the .FR fields depend on each other. I was using it on the state before passing it out at one point, but now I'm not sure - maybe we want to save everything to the database? At any rate, we will need it to handle validation (we don't care about errors in the the organization fields for individuals), but it's not being used right now.  I should probably pull it out, but I don't think I will just yet unless someone objects strongly.

mikeshelton1503 commented:
@deBhal Finally taking a look here. It's looking good, nice work!

I noticed a few things that need some more attention:

- The Section Header for the additional step should read ".FR Registration" (or whatever the ccTLD is in the particular case).
- For .fr, the country of birth field was set to Australia for me. It should either be empty or default to France.
- I know @ranh suggested removing it but I'd still like to have the intro header and paragraph per the design in #13890. We can tweak the copy some more but I do believe that some explanation is necessary. Also as @deBhal suggested it might be helpful to include a link to a support page (though generally I like to avoid links in a purchase flow).

Other than that, I'll work on updating the presentation of the date fields and there are couple other  minor design improvements I'm going to make. I'll work on those changes tomorrow (Thursday).

Also, just a note, I did my best to research what a VAT, SIREN, SIRET, and EU Trademark Number should look like for the placeholder text, which is what you have implemented thus far. However, we should probably have someone more familiar with those numbers review.

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,319 @@
+							name="registrantType"
+							checked={ 'individual' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
> The Section Header for the additional step should read ".FR Registration" (or whatever the ccTLD is in the particular case).
Fixed! Thanks for noticing :)

> For .fr, the country of birth field was set to Australia for me. It should either be empty or default to France.
Done. However, I wonder if we should default to the country code from the first form instead of 'FR'?

> I'd still like to have the intro header and paragraph per the design in #13890. 
Done!

> (though generally I like to avoid links in a purchase flow).
Yeah, that's a really good point. Perhaps some sort of optional docs in-page? I don't want to hold up this PR to solve that, but maybe we can come back to it.

> Also, just a note, I did my best to research what a VAT, SIREN, SIRET, and EU Trademark Number should look like for the placeholder text, which is what you have implemented thus far.

I did double checked the format of all these against the [opensrs docs](http://domains.opensrs.guide/v1.0/docs/tld#section-fr-registrant_extra_info) and wikipedia, and made a few small adjustments (your trademark was missing a 7 (and definitely should be 9 digits), I used more numbers instead of 0s in the SIRET number and I used Paris' post code for the example). I'm pretty confident that they're ok.

ranh commented:
> We can tweak the copy some more but I do believe that some explanation is necessary

I originally suggested we remove this part because I couldn't think of a short helpful message to use here. I felt like no message would be better than an unhelpful one. Without any message, this screen is just "more stuff to fill out". But if we try to explain why we ask for this information, there's much more room for confusion, misunderstanding, and open questions. 

As a side note, I also think we may have a bias "against" this extra screen, because we are so familiar with the shorter flow we have for gTLDs. If we look at ccTLD registration as a complete flow, not as an add-on over an existing flow, then this screen is no longer a new or special, it's just another step in the process. And as such, it does not require more explanation than any other step.

Of course, happy to try to tweak this further. Can you expand on what/why you think we could use more explanation here?

mikeshelton1503 commented:
Perhaps some missing context here is that I also believe the regular domain contact info form needs a description too, similar to how we did on get.blog (just thought that might be best as a separate change). So it's not just about providing context for this new step, it's about guiding the user through the whole flow with some subtle, encouraging copy that briefly captures the purpose of the "step" they're on. No need to over explain or add too much detail (which my original copy probably was guilty of) but with a lengthy stepped process like this I've typically seen positive results from friendly guidance copy. It can help ease the tension of filling out the form.

>Without any message, this screen is just "more stuff to fill out". But if we try to explain why we ask for this information, there's much more room for confusion, misunderstanding, and open questions.

Agreed that how we say this is very important and if we say too much it could lead to additional questions, however I disagree with the assumption that users won't have doubts just by saying nothing. The unknown can cause just as much or more anxiety. 

>"more stuff to fill out"

I do believe that many users, who have purchased other things on the web before do have a tendency to just go with the flow and fill out a form. However, we're collecting more than your standard purchase flow info, even in the "basic" domain contact info form but especially with some of the ccTLDs.

ranh commented:
On get.blog we say "Your details are needed to register example.blog." 

We can try something like this for wpcom.

For the default contact screen:

> Your contact information is needed to register example.com.

For the ccTLD screen:

> We need some extra details to register domains ending in ".fr".

mikeshelton1503 commented:
>However, I wonder if we should default to the country code from the first form instead of 'FR'?

Good call. Let's do that!

deBhal commented:
>For the default contact screen:
>
>>Your contact information is needed to register example.com.

I'm very reluctant to mess with the main page in any way that will effect users without a ccTLD. It's clear the domains page is problematic, and If I have to demonstrate that I haven't made it any worse it will hold up the whole ccTLDs project. 

What about suggesting this on #12622?

> For the ccTLD screen:
>
>> We need some extra details to register domains ending in ".fr".
Nice! Done:

![checkout_ _julesauspremiumtest_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/26476464/53d1e7e8-4202-11e7-9a7d-46ea20d13b34.jpg)

> default to the country code from the first form instead of 'FR'
Done.

mikeshelton1503 commented:
@deBhal No problem we can make the header change separately of this PR.

klimeryk commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -160,11 +198,32 @@ class DomainDetailsForm extends Component {
+		const allFieldValues = Object.assign(
+			{},
+			formState.getAllFieldValues( this.state.form ),
+			this.state.registrantExtraInfo
```
I've noticed that if you don't fill out, for example, your birth date or any extra info, it won't get submitted to the backend for validation at all:
<img width="534" alt="screen shot 2017-05-26 at 13 01 02" src="https://cloud.githubusercontent.com/assets/3392497/26492225/69c84a0e-4213-11e7-80d6-a729ce2afcac.png">
Is this expected?

deBhal replied:
Yes, this is how I decided to go for now, but we won't know if it's right until we get validation and pre-populating-from-the-db working.

I went this way now because:
- What do we do with validation errors for fields that aren't visible?
- There's no meaningful notion of "complete" with extra info anyway, what we care about depends on the domains and on the fields themselves
- We're going to need to do some reasoning on the back end to figure out which fields matter either way
- Backend processing is more plentiful than bandwidth
- There are no dependencies yet between the "core" data and the "extra" data.
- the empty fields will become meaningful when we repopulate the form from the database.

Redux gives us a solid answer to "where do we put the errors" (they go in the state!), so I think that what we want to do is send everything we've got from the user, but not pad it out with empty fields.

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,351 @@
+					<FormLegend>
+						{ translate( 'Date of Birth' ) }
+					</FormLegend>
+					<div className="registrant-extra-info__dob-inputs">
```
How about using https://wpcalypso.wordpress.com/devdocs/design/date-picker instead? It should provide a much better UX (not to mention that for a person in US, the month should be first in the input, etc. ;)) with hopefully less code 😃 
Otherwise, this approach needs more strict validation (you can input more than 4 digits for the year, etc.) - all that should be taken care of by DatePicker.

mikeshelton1503 replied:
Date pickers are only a better UX when choosing a date that you're unfamiliar with (like travel dates) where to context of the days of the week is helpful. Also, it doesn't work well for DOB because its hard (mostly impossible) to navigate to _your_ year of birth.

klimeryk replied:
Ah, good point - I hadn't thought about it that way 🤔 I saw the input fields and thought: there's got to be a better way!

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,351 @@
+						<FormTextInput
+							name="placeOfBirth"
+							value={ placeOfBirth }
+							placeholder={ translate( 'Place or city of birth' ) }
```
As a user, the phrasing "Place _or_ city of birth" could be a bit confusing? What's a valid "place" to put here? Hospital X? (I'm actually confused about this myself 😅)

klimeryk replied:
Looking at the SRS docs, they just mention `The city in which the individual was born.`, so maybe let's keep it simple too?

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,351 @@
+				{ countryOfBirth === 'FR' && (
+					<FormFieldset>
+						<FormLabel>
+							{ translate( 'Postal Code of Birth' ) }
```
Seriously? WTF, postal code of birth? 😩 I guess I should be grateful they don't require you to submit the name of the doctor that delivered you...

deBhal replied:
I know, right!?  On the upside, AFNIC has dropped this, let's just hope OpenSRS does too.

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,351 @@
+				<FormFieldset>
+					<FormLabel className="registrant-extra-info__optional">
+						{ translate( 'VAT Number' ) }
+						{ this.renderOptional() }
```
Just wanted to note that it looks very nice 👍

klimeryk commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -330,6 +397,18 @@ class DomainDetailsForm extends Component {
 
+	renderExtraDetailsForm = () => {
+		return ( <ExtraInfoFrForm
+			isProbablyOrganization={ Boolean( formState.getFieldValue( this.state.form, 'organization' ) ) }
```
Nice UX! 👍

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,354 @@
+							id="registrantType"
+							checked={ 'individual' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
I've put in all the `for` attributes, thanks for pointing it out!

I spent a little too long trying to run it through some testing tools without any success, so if you could check they're right I'd appreciate it. In particular, do we need a `htmlFor` for the radio buttons inside the labels? The spec says no, but I saw a comment about older browsers needing one anyway, but didn't find anything more specific.

(P.s. I wrote linting rules for my pet peeves, and life is better ([e.g.](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/i18n-no-collapsible-whitespace.md)). I'm not going to stop and do it right now, but I'd love to help or to look into it after the growth rotation. The readme for [eslint-plugin-jsx-a11y/label-has-for.md](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/label-has-for.md) looks promising, and pulling in an accesibility linting library seems double-plus good! Actually, please don't do that without me :) )

I've also fixed the validation check before progressing - great catch! I was worried this was going to be really fiddly, but turns out I just missed the right spot by a half dozen lines :) The step progression is inside the validation check now, so it waits for validations and aborts if it fails:
![fr extra validate before you go go](https://cloud.githubusercontent.com/assets/5952255/26565642/43ae9e5c-4530-11e7-989c-611aa3bc93d4.gif)

I've also fixed up the case and copy as you suggested and responded to your question about what we send to the back end (short version: it's doing what I intended, we'll see if I'm right).

The two things I haven't addressed:

1) The DOB Picker - We discussed this a bit on the issue https://github.com/Automattic/wp-calypso/issues/13890#issuecomment-300653680 with some background links, and the problem with the date picker is that it's poorly suited to picking a birth year, so we do need something more specific. You're totally right about needing better validation, and you can see I've had a look at the logic, but:
2) Validation in general: This is important, but it will be easier and better later.  Yoav is working on back end validation and Hew is working on reduxifying the contact details data, so I want to punt validation/sanitation until those are in (particularly the redux - validating against the global seems so much simpler than working around the components). We definitely can't turn on the feature without it, though, so I've stuck some "FIXME" comments in there as well.

Are you happy with all of that?

deBhal commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -92,14 +110,34 @@ class DomainDetailsForm extends Component {
+		return this.state.currentStep !== first( this.state.steps );
+	}
+
+	switchToNextStep = () => {
```
✅

mikeshelton1503 commented:
>I was wondering if there should be some sort of transition effect between these two contact forms - we show an effect when showing this form, and the second step is so quick to appear that the user might be confused? I don't know if that won't be too distracting, so probably just something for @mikeshelton1503 to shoot down ;)

No shooting here, I think this is a good suggestion. It's a bit jarring without it sense we use the transition elsewhere.

a8ci18n commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -424,7 +518,10 @@ class DomainDetailsForm extends Component {
-		if ( needsOnlyGoogleAppsDetails ) {
+		// TODO: gather up tld specific stuff
+		if ( this.state.currentStep === 'fr' ) {
+			title = this.props.translate( '.FR Registration' );
```
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal replied:
👎

deBhal commented:
> I was wondering if there should be some sort of transition effect

Done! That was a lot more fiddly than I expected, I hope I did it the right way. I wanted to apply the animation to the top level payment container, but activate it based on the different forms, so I re-wrote the animation as a `ReactCSSTransitionGroup`, and I put that inside the `PaymentBox`, which meant I also need to upgrade that to es6 to keep the linter happy.

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,369 @@
+							<FormTextInput className="registrant-extra-info__dob-year"
+								id="dobYears"
+								value={ dobYears }
+								max="4"
```
Good call on `type='number'`, but you probably wanted to use `maxLength`, not `max` (maximum value) :) ...but after testing, it seems that doesn't work either. It looks like `maxlength` is [not supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number) for `type=number` inputs :man_facepalming:

deBhal replied:
I'll just drop the `max` for now, but css validation does seem like bit of a missed opportunity.

I don't know if our custom submit handling is compatible with the validation check, but even if it wasn't we could have the error elements ready and then make them visible for invalid values.

Js validation seems a fair choice, though. Much more flexibility and control.

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,369 @@
+					</FormLegend>
+					<div className="registrant-extra-info__dob-inputs">
+						<div className="registrant-extra-info__dob-column">
+							<FormLabel htmlFor="dobYears" className={ screenReaderText }>
```
Nice! I was wondering how to use this class after you've asked about it on Slack :) TIL!

klimeryk commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,369 @@
+								value={ dobYears }
+								max="4"
+								type="number"
+								placeholder="YYYY"
```
I think the placeholders for these DOB fields should probably be `translate`d too - everyone might get the `YYYY` is for `Year`.

deBhal replied:
Oh, good catch! Bonus points for schooling me on i18n. I don't know if that's a :), :(, :/ or :'(. Maybe a bit of each.

klimeryk replied:
![1q2u3l](https://cloud.githubusercontent.com/assets/3392497/26660555/1479ad08-4679-11e7-9039-3f3e170412bc.jpg)

deBhal commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,351 @@
+	return [ 'registrantType', ...conditionalFields ];
+}
+
+class RegistrantExtraInfoForm extends React.PureComponent {
```
✅

deBhal commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,369 @@
+	}
+}
+
+RegistrantExtraInfoForm.displayName = 'ExtraInfoFrForm';
```
Oh, neat, TIL!

That wasn't true a few weeks ago - the minimized code meant that devdocs showed things like `r` for the title :)

Do you have a link handy?

klimeryk replied:
I haven't seen any internal announcements about this, but it was part of this massive PR for upgrading to Babel 6 long time ago (Dec 2015): https://github.com/Automattic/wp-calypso/pull/1515. Official docs: https://babeljs.io/docs/plugins/transform-react-display-name/ (not very helpful :/).

I wonder why it did not work in that case. Maybe there was no explicit variable involved?
```
// Will _not_ work:
export default class X extends Component { }

// Works:
class X extends Component {}
export default X;
```

deBhal commented on client/components/domains/registrant-extra-info/docs/example.jsx:
```diff
@@ -0,0 +1,48 @@
+	constructor( props ) {
+		super( props );
+
+		this.state = {
```
Awesome, Thanks! I tried a static `initialState` with no joy, but didn't take that next logical step (nor go and read the manual) :p

deBhal replied:
FYI: no `static`, it's just:
```
	state = {
		whatever: 'value',
	}
```
Which does make sense (we want the value on the object not the prototype), but it's kind of weird syntax to be just running code there (I stuck a `console.log()` in there to check, and it is just running that code before the constructor with the instance as `this`).

Thinking about it though, this is the same thing as the binding.

klimeryk replied:
Bah, sorry about that - wrote that from memory without thinking :) Of course, it should not be `static` :man_facepalming:

klimeryk replied:
Unless I'm missing something - it's not a weird syntax :) It's so normal that it's weird 😀 You're doing the same thing you're doing in the constructor - assigning an instance variable `state` some value.

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,352 @@
+							id="registrantType"
+							checked={ 'individual' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -424,7 +505,10 @@ class DomainDetailsForm extends Component {
-		if ( needsOnlyGoogleAppsDetails ) {
+		// TODO: gather up tld specific stuff
+		if ( this.state.currentStep === 'fr' ) {
+			title = this.props.translate( '.FR Registration' );
```
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Registration' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/tld/fr-extra-fields-dialog)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -347,7 +427,7 @@ class DomainDetailsForm extends Component {
 
 	handleSubmitButtonClick = ( event ) => {
-		event.preventDefault();
+		event && event.preventDefault();
```
No, just vestigial. I was experimenting with passing pseudo submit events shortly before I threw my hands in the air and said "Reduxify!" :)

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,352 @@
+							id="registrantType"
+							checked={ 'individual' === registrantType }
+							onChange={ this.handleChangeEvent } />
+						<span>{ translate( 'An individual' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **7** times:
`translate( 'Individual' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented on client/my-sites/upgrades/checkout/domain-details-form.jsx:
```diff
@@ -30,11 +34,16 @@ import { countries } from 'components/phone-input/data';
 	countriesList = countriesListForDomainRegistrations();
 
-class DomainDetailsForm extends Component {
+class DomainDetailsForm extends PureComponent {
```
✅ 

👍  for adding more information with this sort of subtlety.

I got surprised a couple of times and overreacted by binding everything with a `this`, but I agree with you - more meaning is good.  I think I've got it now: do bind anything you 
 pass to other components (e.g. event handlers), don't bind react lifecycle methods. If something complains about a missing this, bind that too ;)

a8ci18n commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -0,0 +1,352 @@
+								value={ dobMonths }
+								max="2"
+								type="number"
+								placeholder={ translate( 'MM', {
```
Hi! I've found a possible matching string that has already been translated **11** times:
`translate( 'mm', { context: 'millimeters'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
@klimeryk: I've implemented that final round of feedback and then squashed for the merge, which makes having a quick look pretty hard, so I've pushed `add/tld/fr-extra-fields-dialog-backup-history` [commits](https://github.com/Automattic/wp-calypso/commits/add/tld/fr-extra-fields-dialog-backup-history). No point leaving it lying around, though, so I'll delete it tomorrow if you don't do it first :)

Bullet points Fixes:
- method binding
- example es6
- placeholder localization

I've also dropped the unused code (into a local branch), the unnecessary check and the `max` and `displayName`.

Thanks again for such a detailed review!

I've done a final check of everything, and I'm going to merge now.

klimeryk commented on client/components/domains/registrant-extra-info/docs/example.jsx:
```diff
@@ -0,0 +1,43 @@
+ */
+import Card from 'components/card';
+import ExtraInfoFrForm from 'components/domains/registrant-extra-info/fr-form';
+import { forDomainRegistrations as getCountries } from 'lib/countries-list';
```
Nice idea!

klimeryk commented:
Congrats on merging this 😀 I've looked through the final version and it looked good - thanks for bearing with me! 🙇

deBhal commented:
> thanks for bearing with me! 🙇

I think you might have that backwards :) I really appreciate such a thoughtful review.
